lib/charms/opensearch/v0/opensearch_peer_clusters.py

```diff
@@ -0,0 +1,374 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Class for Managing simple or large deployments and configuration related changes."""
+import logging
+from typing import List, Optional
+
+import shortuuid
+from charms.opensearch.v0.constants_charm import (
+    CMRoleRemovalForbidden,
+    CmVoRolesProvidedInvalid,
+    DataRoleRemovalForbidden,
+    PClusterNoRelation,
+    PClusterWrongNodesCountForQuorum,
+    PClusterWrongRelation,
+    PClusterWrongRolesProvided,
+    PeerClusterRelationName,
+    PeerRelationName,
+)
+from charms.opensearch.v0.helper_cluster import ClusterTopology
+from charms.opensearch.v0.models import (
+    DeploymentDescription,
+    DeploymentState,
+    DeploymentType,
+    Directive,
+    Node,
+    PeerClusterConfig,
+    StartMode,
+    State,
+)
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchError
+from charms.opensearch.v0.opensearch_internal_data import RelationDataStore, Scope
+from ops import BlockedStatus
+
+# The unique Charmhub library identifier, never change it
+LIBID = "35ccf1a7eac946ec8f962c21401598d6"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchProvidedRolesException(OpenSearchError):
+    """Exception class for events when the user provided node roles will violate quorum."""
+
+
+class OpenSearchPeerClustersManager:
+    """This class covers the configuration changes depending on certain actions."""
+
+    def __init__(self, charm):
```

Comment:
**(strong) suggestion** I would in general avoid to provide the charm as argument to this. I understand this is fairly common, but this really creates a bit of circular  dependency here, if we were to type this, since the Charm needs to import `OpenSearchPeerClustersManager`, that however would need to import the charm type. 

I suppose eventually this class should have knowledge of:
1. Config of the charm, represented by the pydantic object
2. Intra-application peer relation (since needs to write/read the `DeploymentDescrition`)
3. Extra-application relation (since the PeerClusterRelation will eventually be implemented)

I noticed that `_opensearch` is only used to have the `is_node_up` method. Maybe instead of passing this object, can just be easier to provide either a boolean or inject a lambda function that does that query, but probably the entire object is really not needed, and created a bit of tricky dependency tree

I understand that the other classes are implemented in this way, but (if you agree with these points) I'd rather start addressing this when creating new classes, than addressing this in a single refactoring, becuase that refactoring would be huge and probably never feasible. I would attach this bit by bit instead...

Reply:
Yes, but just to be clear, what suggested by Marc is just a workaround to type things, but it does not address the circular dependency issue, which - in general - makes also components way more tightly connected one another, reducing/limiting modularity, testability and separation of concerns. Preveting the typing (as addressed by the workaround) is only one of the drawbacks, but not the only one 

I'd argue circular dependency is a bit of an anti-pattern in software engineering, which I would in general encourage to avoid whenever possible

Reply:
anyhow, I don't think it is worth to stop the PR if this requires substantial refactoring outside the scope of this PR